### ISSUE TITLE
Dashboard: Missing DNS Rebinding attack protection

### ISSUE THREAD

**Issue by admc** [linkerd/linkerd2#3083]
From our external security audit.

### What is the issue?

It was found that the Linkerd dashboard service, as well as the Grafana dashboard, are vulnerable to a DNS Rebinding attack. This allows an attacker to have a victim visiting an attacker-controlled website. The attacker will then be able to perform requests and retrieve responses on the victim’s behalf.

The attack exploits the fact that the dashboard does not restrict the Host header for incoming requests. By setting up a script on a domain with a very low TTL value and then quickly rebinding its IP address of localhost, the script will then “bypass” the restrictions normally set by SOP on the browser.

### How can it be reproduced?

- Run Linkerd dashboard service with port 3000 (linkerd dashboard --port 3000)
- Go to http://rebind.it:8080/manager.html and change the target port to 3000.
- Click on “Start Attack” and wait for a minute.
- The response from the dashboard will be returned, indicating success of a DNS rebinding attack

### Additional context

It is recommended to provide an option to fixate the dashboard domain. Grafana, for example, provides an enforce_domain option to ensure that redirection to the correct domain happens consistently.

**Comment by olix0r** [comment]
It seems that this is complicated by our port-forwarding logic: clients may choose arbitrary ports for the api service. We must allow for an arbitrary number of ports, as clients may wish to connect to multiple cluster dashboards simultaneously.

If we wanted to hardcode a check in the web service, we could test that the host header matched `127.0.0.1:\d+`, etc; or, we could determine a fixed port range that we serve on (say, 100 ports) to limit the attack surface area.

But even if we do this, it's not going to be possible to configure grafana to participate in such a scheme, since its configuration only supports a single fixed hostname.

Generally, it seems that it would be a great feature for Linkerd to enforce in the mesh itself -- that pods only receive requests to a blessed set of service names. This would help to provide DNS rebinding protection at the infrastructure level! But, this is complicated by the way in which [kubectl port-forwards skip the linkerd proxy](https://github.com/linkerd/linkerd2/issues/2352), so our dashboard is typically exposed without the benefit an inbound proxy to enforce this sort of policy.

In summary, in order to remediate this -- whether in the mesh or via a frontend HTTP router for the control plane -- we're likely going to have to alter how the web service is exposed to clients via the command-line.

**Comment by olix0r** [comment]
It would be fairly straightforward to address this in the `dashboard` command itself if we exposed a local HTTP server that enforced the `localhost:PORT` name requirements before forwarding requests through the port-forward.

**Comment by olix0r** [comment]
To elaborate, I'd like to propose the following fix:

1. Modify the `dashboard` command to wrap the exposed forwarded socket with an HTTP middleware that (1) returns 404s if a request includes an unexpected host header and (2) rewrites the forwarded request with a host header as described below.
2. Modify linkerd-web to return 404 for requests that are not to `linkerd-web.${linkerd_ns}.svc.${cluster_name}:8400`
3. Modify grafana to use `ensure_domain linkerd-grafana.${linkerd_ns}.svc.${cluster_name}:3000`.
4. Update documentation to describe how to expose the linkerd dashboard via an ingress (i.e. requiring header rewrites).

However, this may not work if the client-go APIs don't give us enough leeway to install an HTTP middlware around the port-forwarded socket...

**Comment by olix0r** [comment]
If we are unable to wrap the port-forward sockets with an HTTP middleware directly, then if the control plane service enforce name requirements, it would be safe to open a web server on localhost that rewrites requests and forewords requests through the local portforward.

**Comment by grampelberg** [comment]
> Modify linkerd-web to return 404 for requests that are not to linkerd-web.${linkerd_ns}.svc.${cluster_name}:8400

This won't work for folks exposing the dashboard externally =/ We'll have to make it either configurable directly or read from ingress objects.

**Comment by alpeb** [comment]
> - Modify linkerd-web to return 404 for requests that are not to linkerd-web.${linkerd_ns}.svc.${cluster_name}:8400
> - Modify grafana to use ensure_domain linkerd-grafana.${linkerd_ns}.svc.${cluster_name}:3000.

@ver While working on this I realized that once we protect `linkerd-web` we won't have to do anything with Grafana because it's being proxied through `linkerd-web`. So Grafana could still be exposed directly or through ingress without any extra config. Agreed?
